src/batch.js

```diff
@@ -0,0 +1,54 @@
+const call    = require('ramda/src/call')
+const curry   = require('ramda/src/curry')
+const juxt    = require('ramda/src/juxt')
+const max     = require('ramda/src/max')
+const zipWith = require('ramda/src/zipWith')
+
+// batch :: { k: v } -> ([a] -> Promise [b]) -> a -> Promise b
+const batch = (opts={}, f) => {
+  const {
+    limit = Infinity,
+    wait  = 32
+  } = opts
+
+  let args     = []
+  let last     = 0
+  let rejects  = []
+  let resolves = []
+  let timeout  = 0
+
+  const batched = arg => {
+    const promise = new Promise((res, rej) => {
+      args.push(arg)
+      rejects.push(rej)
+      resolves.push(res)
+    })
+
+    if (args.length >= limit) run()
+
+    else if (!timeout) {
+      const delta = new Date() - last
+      timeout = setTimeout(run, max(0, wait - delta))
+    }
+
+    return promise
+  }
+
+  const run = () => {
+    Promise.resolve(args)
+      .then(f)
```

Comment:
Does this assume that `f` will return an array that has the same length of args & will be in the same order? (i.e, the result of `args[0]` is `fn(args)[0]`, `args[1]` is `fn(args)[1]`, etc?

Reply:
Makes sense.

(I'm just rambling... don't want to hold you up.)

The use case i'm thinking of is a function that returns a key/value object. Hoping for an easy win.

```javascript
// getProfile :: [ id, ... ] -> Promise { [id]: obj, ... }

const getProfileBatched = batch(options, getProfile)

map(evolve({ profile: getProfileBatched }), data)
```

Reply:
Yeah, I can see that.  In your case the args are strings, which are easy to match against.  But in the general case, they could be any value, including objects or arrays, neither of which are easy to match against, so the easiest way is to maintain the ordering of inputs-to-outputs.

Reply:
Could use a `Map`, i guess.